### ISSUE TITLE
Missing Relay Ice Candidate

### ISSUE THREAD

**Issue by eirikhollis** [microsoft/MixedReality-WebRTC#42]
Hi,

I'm currently moving from an implementation built around the m71 release of webrtc-uwp-sdk to this one, and have stumbled upon an problem I'm not able to resolve. The solution works on webrtc-uwp-sdk.

I have a TURN/STUN Kurento server running on Azure, and when I connnect to it (or any peers through it) with the previous solution, my client always sent 5 ice candidates.
4 local and 1 relay. The relay one looks like this:

`candidate:[number] 1 udp [priority] [Server IP] [WebRTC port] typ relay raddr [Local IP] rport [WebRTC port] generation 0 ufrag 9dPY network-id 1 network-cost 50`

However, on this plugin, the relay candidate is never generated (or at least never given to me). I only receive the 4 local ones, and therefore can't setup a connection through the server.
I have added the ice servers to the peer as

```
iceServers = new List<string>
{
      $"stun: {url}{port}",
      $"turn: {url}{port}",
      $"turn: {url}{port}?transport=tcp"
};
```
and have also added Username and Credential for the TURN server.

![image](https://user-images.githubusercontent.com/30655189/63753599-a84f9600-c8b3-11e9-9d37-0b85c225fd97.png)

Within the same environment, same computer/network/external peer, I can connect with the old solution both before and after trying this one.

Am I missing anything? Shouldn't this then generate the relay candidate? Or is the plugin somehow configured for local communication only?

While on the topic, would be amazing if we have access to modifying the transport- and bundle-policy of ice gathering.

**Comment by eirikhollis** [comment]
Solved it.
Often typing out an issue helps resolving it.

The issue was the space in the ice server list, between the type identifier and the url. Correct implementation is
```
iceServers = new List<string>
{
      $"stun:{url}{port}",
      $"turn:{url}{port}",
      $"turn:{url}{port}?transport=tcp"
};
```
I followed the design in the PeerConnection.cs script, which has the ToString method:
```
public override string ToString()
{
     return string.Format("{0}: {1}", Type.ToString().ToLower(), Uri);
}
```
Should the space probably be removed here as well?

**Comment by djee-ms** [comment]
Hi @eirikhollis

Thanks for all the details. Yes it makes no sense to have a space here between the server URL and port. And in fact the default value in the Unity integration doesn't have a space:
https://github.com/microsoft/MixedReality-WebRTC/blob/12b1c55e318191cc0462048d4008a2aafaa18960/libs/Microsoft.MixedReality.WebRTC.Unity/Assets/Microsoft.MixedReality.WebRTC.Unity/Scripts/PeerConnection.cs#L151

I will fix the `ToString()` method. Sorry about that.

**Comment by djee-ms** [comment]
Confirmed with the source implementation, there is no space in the scheme:
`turnURI = scheme ":" turn-host [ ":" turn-port ] [ "?transport=" transport ]`
https://github.com/webrtc-uwp/webrtc/blob/baa89682df1ab891ae274a1cb2d9be3b1ee7e20b/pc/iceserverparsing.cc#L128-L148

**Comment by eirikhollis** [comment]
Makes sense. Thanks for the quick reply!

**Comment by djee-ms** [comment]
Fixed with 573c04b06c2b0ea2ab95ebbbf99d2bce1d23077d. Closing this issue, thanks for the feedback!

**Comment by michael-house** [comment]
I was apparently using the version of this code that included the space. It took a while to track down!

It would be nice if when adding ICE Servers, there was a check for malformed URLs. Since the only feedback is the successful generation of local ICE candidates. It's not very clear that there's an error with the ICE candidate generation, unless you know in advance exactly what to expect like eirikhollis. Reporting issues connecting to a STUN/TURN server would be ideal, but at least checking the URLs as they are added to the peer would be nice.
